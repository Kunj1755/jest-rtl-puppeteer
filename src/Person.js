import React, { useState } from "react";
const person = props => {
    const [count, setCount] = useState(0);
    const _onClick = () => {
        setCount(count + 1);
    }
    return (
        <div>
            <p onClick={_onClick}>
                I'm a person and My name is {props.name} and I am {props.age} years old
      </p>
            <p> {props.children}</p>
            <p >{count}</p>
            <p id="para">100</p>
            <input type="text" onChange={props.changed} />
        </div>
    );
};
export default person;
