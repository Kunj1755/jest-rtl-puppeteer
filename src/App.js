import React, { Component } from "react";
import Person from "./Person";

import "./styles.css";

class App extends Component {
  state = {
    persons: [
      { id: "dudu", name: "Ram", age: "20" },
      { id: "wjow", name: "Shyam", age: "25" },
      { id: "ywey", name: "Mohan", age: "28" }
    ],
    otherState: "Default state",
    showPersons: false
  };

  switchNameHandler = newName => {
    console.log(newName);
    this.setState({
      persons: [
        { name: newName, age: "20" },
        { name: "Shyam", age: "25" },
        { name: "Mohan", age: "29" }
      ]
    });
  };
  // The event object is automatically passed by JavaScript.
  nameChangedHandler = (event, id) => {
    // We want to update the state of the person but only for one into which input field we typed
    const personIndex = this.state.persons.findIndex(p => {
      return p.id === id;
    });

    // Not a good practice to mutate state directly. A person in persons array is a javascript object. JavaScript objects are reference types. So we should not mutate them directly bcz we only get a pointer when we reach put to person in the next statement. Hence we would mutate the original object to which a pointer points.
    //const person = this.state.persons[personIndex];

    // Better approach (ES6)
    const person = {
      ...this.state.persons[personIndex]
    };
    // Alternative approach
    //const person = Object.assign({}, this.state.persons[personIndex]);

    // Manipulating the copied object, not the original one
    person.name = event.target.value;

    const persons = [...this.state.persons];
    persons[personIndex] = person;

    this.setState({ persons: persons });
  };

  render() {
    return (
      <div className="App">
        <h1> Hello React!</h1>
        <button onClick={this.switchNameHandler.bind(this, "Ramavatar")}>
          Switch name
        </button>
        <Person
          name={this.state.persons[0].name}
          age={this.state.persons[0].age}
        />
        {/* Add a property 'click' and pass ref to handler click={this.SwitchNameHandler} */}
        <Person
          name={this.state.persons[1].name}
          age={this.state.persons[1].age}
          click={this.switchNameHandler.bind(this, "Ram Krishna")}
          changed={event =>
            this.nameChangedHandler(event, this.state.persons[1].id)
          }
        >
          My Hobby : TT
        </Person>
        <Person
          name={this.state.persons[2].name}
          age={this.state.persons[2].age}
        />
      </div>
    );
  }
}

export default App;
