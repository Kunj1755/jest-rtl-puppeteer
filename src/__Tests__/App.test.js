import React from "react";
import TestRenderer from "react-test-renderer";
import { cleanup, fireEvent, render } from '@testing-library/react';
import App from "../App";
import Person from "../Person";

// automatically unmount and cleanup DOM after the test is finished.
afterEach(cleanup);

// const personCompProps = {
//     name: "Ram",
//     age: "30",
//     handleOnClick: jest.fn(),
//     handleOnChange: jest.fn()

// };

test("App renders without crashing", () => {
    const testRenderer = TestRenderer.create(<App />);
    const testInstance = testRenderer.root;
    // console.log(testRenderer);
    // console.log(testInstance);
    expect(testInstance.findByType(Person).props.name).toBe('Shyam');
    expect(testInstance.findByType(Person).props.age).toBe('25');
    expect(testInstance.findByProps({ id: "para" }).children).toEqual(['100']);
    const tree = testRenderer.toJSON();
    expect(tree).toMatchSnapshot();
});

