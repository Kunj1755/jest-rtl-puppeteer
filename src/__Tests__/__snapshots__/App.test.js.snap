// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders without crashing 1`] = `
<div
  className="App"
>
  <h1>
     Hello React!
  </h1>
  <button
    onClick={[Function]}
  >
    Switch name
  </button>
  <div>
    <p
      onClick={[Function]}
    >
      I'm a person and My name is 
      Shyam
       and I am 
      25
       years old
    </p>
    <p>
       
      My Hobby : TT
    </p>
    <p>
      0
    </p>
    <p
      id="para"
    >
      100
    </p>
    <input
      onChange={[Function]}
      type="text"
    />
  </div>
</div>
`;
